compiler/rustc_hir_analysis/src/astconv/mod.rs

```diff
@@ -2098,15 +2095,157 @@ impl<'o, 'tcx> dyn AstConv<'tcx> + 'o {
         Ok((ty, DefKind::AssocTy, assoc_ty_did))
     }
 
+    fn lookup_inherent_assoc_ty(
+        &self,
+        name: Ident,
+        segment: &hir::PathSegment<'_>,
+        adt_did: DefId,
+        self_ty: Ty<'tcx>,
+        block: hir::HirId,
+        span: Span,
+    ) -> Result<Option<(Ty<'tcx>, DefId)>, ErrorGuaranteed> {
+        let tcx = self.tcx();
+
+        let candidates: Vec<_> = tcx
+            .inherent_impls(adt_did)
+            .iter()
+            .filter_map(|&impl_| Some((impl_, self.lookup_assoc_ty_raw(name, block, impl_)?)))
+            .collect();
+
+        if candidates.is_empty() {
+            return Ok(None);
+        }
+
+        let cause = traits::ObligationCause::misc(span, block);
+        let mut unsatisfied_predicates = Vec::new();
+
+        for &(impl_, (assoc_item, def_scope)) in &candidates {
+            let infcx = tcx.infer_ctxt().ignoring_regions().build();
+            let param_env = tcx.param_env(impl_);
+
+            let impl_ty = tcx.bound_type_of(impl_);
+            let impl_substs = self.fresh_item_substs(impl_, &infcx);
+            let impl_ty = impl_ty.subst(tcx, impl_substs);
+
+            let InferOk { value: impl_ty, obligations } =
+                infcx.at(&cause, param_env).normalize(impl_ty);
+
+            // Check that the Self-types can be related.
+            let Ok(InferOk { obligations: sub_obligations, value: () }) = infcx
+                .at(&ObligationCause::dummy(), param_env)
+                .define_opaque_types(false)
+                .sup(impl_ty, self_ty)
```

Comment:
Does method selection use `sup`? We probably should use `eq` here, since that's what [`match_impl`](https://doc.rust-lang.org/nightly/nightly-rustc/src/rustc_trait_selection/traits/select/mod.rs.html#2406) does.

Reply:
Well, method *probing* does use `sup` but tbh I am not sure why exactly since it is `.ignoring_regions()`, too, and lifetime subtyping is the only subtyping in Rust's type system. I can definitely switch to `eq` but I wish I could understand for what deeper reason the probing code uses `sup`.

https://github.com/rust-lang/rust/blob/222d1ff68d5bfe1dc2d7f3f0c42811fe12964af9/compiler/rustc_hir_typeck/src/method/probe.rs#L1516

You likely know the following, I am just making sure I understand correctly. As far as I know method probing ignores regions since implementation-wise it uses `FnCtxt` which stores [`Inherited`](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_hir_typeck/struct.Inherited.html) whose only public ctor [`Inherited::build`](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_hir_typeck/struct.Inherited.html#method.build) constructs an `InferCtxt` `.ignoring_regions()` and since specification-wise regions are erased before codegen and they should matter for method dispatch.

Reply:
If I replace `.sup` with `.eq` here, it ices.

<details><summary>backtrace</summary>

```
thread 'rustc' panicked at 'index out of bounds: the len is 0 but the index is 0', compiler/rustc_infer/src/infer/region_constraints/mod.rs:459:9
stack backtrace:
   0:     0x7f6d278ebac1 - <std::sys_common::backtrace::_print::DisplayBacktrace as core::fmt::Display>::fmt::hef7c34ced48956ed
   1:     0x7f6d27967fbe - core::fmt::write::hd6821393d61dbc78
   2:     0x7f6d27910171 - std::io::Write::write_fmt::h0aae2c130bb43485
   3:     0x7f6d278eb8ea - std::sys_common::backtrace::print::hefc1376ab8e30bd1
   4:     0x7f6d278eca67 - std::panicking::default_hook::{{closure}}::h133538e0d38da3d4
   5:     0x7f6d278ec895 - std::panicking::default_hook::h0f8bde92f9834261
   6:     0x7f6d282873c5 - rustc_driver[748794e2e144500f]::DEFAULT_HOOK::{closure#0}::{closure#0}
   7:     0x7f6d278ecf73 - std::panicking::rust_panic_with_hook::h6563697fa9ccd808
   8:     0x7f6d278e3089 - std::panicking::begin_panic_handler::{{closure}}::ha90000309b42d4bb
   9:     0x7f6d278e2fbc - std::sys_common::backtrace::__rust_end_short_backtrace::h13266d7b6e524bca
  10:     0x7f6d278ecb3a - rust_begin_unwind
  11:     0x7f6d278ca4e3 - core::panicking::panic_fmt::hee5867c2cd98c87f
  12:     0x7f6d278ca652 - core::panicking::panic_bounds_check::h879d026d129b4b6a
  13:     0x7f6d2a4d1938 - <rustc_infer[c33f17b4b4dc7e29]::infer::region_constraints::RegionConstraintCollector>::universe
  14:     0x7f6d2a5032bb - <rustc_infer[c33f17b4b4dc7e29]::infer::InferCtxt>::universe_of_region
  15:     0x7f6d2a43394a - <rustc_infer[c33f17b4b4dc7e29]::infer::combine::Generalizer as rustc_middle[e87e357ae1384561]::ty::relate::TypeRelation>::regions
  16:     0x7f6d2a43bca2 - rustc_middle[e87e357ae1384561]::ty::relate::super_relate_tys::<rustc_infer[c33f17b4b4dc7e29]::infer::combine::Generalizer>
  17:     0x7f6d2a4330cd - <rustc_infer[c33f17b4b4dc7e29]::infer::combine::Generalizer as rustc_middle[e87e357ae1384561]::ty::relate::TypeRelation>::tys
  18:     0x7f6d2a434592 - <rustc_infer[c33f17b4b4dc7e29]::infer::combine::CombineFields>::instantiate
  19:     0x7f6d2a47c80f - <rustc_infer[c33f17b4b4dc7e29]::infer::equate::Equate as rustc_middle[e87e357ae1384561]::ty::relate::TypeRelation>::tys
  20:     0x7f6d2a5511f6 - <rustc_middle[e87e357ae1384561]::ty::subst::GenericArg as rustc_middle[e87e357ae1384561]::ty::relate::Relate>::relate::<rustc_infer[c33f17b4b4dc7e29]::infer::equate::Equate>
  21:     0x7f6d2a57204c - <core[a759e9f1dec68147]::result::Result<rustc_middle[e87e357ae1384561]::ty::subst::GenericArg, rustc_middle[e87e357ae1384561]::ty::error::TypeError> as rustc_type_ir[98cc6c8502f90b82]::InternIteratorElement<rustc_middle[e87e357ae1384561]::ty::subst::GenericArg, &rustc_middle[e87e357ae1384561]::ty::list::List<rustc_middle[e87e357ae1384561]::ty::subst::GenericArg>>>::intern_with::<core[a759e9f1dec68147]::iter::adapters::map::Map<core[a759e9f1dec68147]::iter::adapters::zip::Zip<core[a759e9f1dec68147]::iter::adapters::copied::Copied<core[a759e9f1dec68147]::slice::iter::Iter<rustc_middle[e87e357ae1384561]::ty::subst::GenericArg>>, core[a759e9f1dec68147]::iter::adapters::copied::Copied<core[a759e9f1dec68147]::slice::iter::Iter<rustc_middle[e87e357ae1384561]::ty::subst::GenericArg>>>, rustc_middle[e87e357ae1384561]::ty::relate::relate_substs<rustc_infer[c33f17b4b4dc7e29]::infer::equate::Equate>::{closure#0}>, <rustc_middle[e87e357ae1384561]::ty::context::TyCtxt>::mk_substs<core[a759e9f1dec68147]::iter::adapters::map::Map<core[a759e9f1dec68147]::iter::adapters::zip::Zip<core[a759e9f1dec68147]::iter::adapters::copied::Copied<core[a759e9f1dec68147]::slice::iter::Iter<rustc_middle[e87e357ae1384561]::ty::subst::GenericArg>>, core[a759e9f1dec68147]::iter::adapters::copied::Copied<core[a759e9f1dec68147]::slice::iter::Iter<rustc_middle[e87e357ae1384561]::ty::subst::GenericArg>>>, rustc_middle[e87e357ae1384561]::ty::relate::relate_substs<rustc_infer[c33f17b4b4dc7e29]::infer::equate::Equate>::{closure#0}>>::{closure#0}>
  22:     0x7f6d2a47bac6 - <rustc_infer[c33f17b4b4dc7e29]::infer::equate::Equate as rustc_middle[e87e357ae1384561]::ty::relate::TypeRelation>::relate_item_substs
  23:     0x7f6d2a43af43 - rustc_middle[e87e357ae1384561]::ty::relate::super_relate_tys::<rustc_infer[c33f17b4b4dc7e29]::infer::equate::Equate>
  24:     0x7f6d2a510fc6 - <rustc_infer[c33f17b4b4dc7e29]::infer::InferCtxt>::super_combine_tys::<rustc_infer[c33f17b4b4dc7e29]::infer::equate::Equate>
  25:     0x7f6d2a47c5ce - <rustc_infer[c33f17b4b4dc7e29]::infer::equate::Equate as rustc_middle[e87e357ae1384561]::ty::relate::TypeRelation>::tys
  26:     0x7f6d2889837a - <rustc_infer[c33f17b4b4dc7e29]::infer::InferCtxt>::commit_if_ok::<rustc_infer[c33f17b4b4dc7e29]::infer::InferOk<()>, rustc_middle[e87e357ae1384561]::ty::error::TypeError, <rustc_infer[c33f17b4b4dc7e29]::infer::at::Trace>::eq<rustc_middle[e87e357ae1384561]::ty::Ty>::{closure#0}>
  27:     0x7f6d289d2db1 - <rustc_infer[c33f17b4b4dc7e29]::infer::at::At>::eq::<rustc_middle[e87e357ae1384561]::ty::Ty>
  28:     0x7f6d2894a509 - <dyn rustc_hir_analysis[2dedc18b32354586]::astconv::AstConv>::associated_path_to_ty::{closure#0}
  29:     0x7f6d289499fa - <dyn rustc_hir_analysis[2dedc18b32354586]::astconv::AstConv>::associated_path_to_ty
  30:     0x7f6d2894e1e8 - <dyn rustc_hir_analysis[2dedc18b32354586]::astconv::AstConv>::ast_ty_to_ty_inner
  31:     0x7f6d28657fcf - <rustc_hir_typeck[5384bad6713105a9]::fn_ctxt::FnCtxt>::to_ty
  32:     0x7f6d285febcf - <rustc_hir_typeck[5384bad6713105a9]::gather_locals::GatherLocalsVisitor>::declare
  33:     0x7f6d287c41b3 - rustc_hir[cf8f75e2b7a3e0eb]::intravisit::walk_block::<rustc_hir_typeck[5384bad6713105a9]::gather_locals::GatherLocalsVisitor>
  34:     0x7f6d286ec69d - rustc_hir_typeck[5384bad6713105a9]::check::check_fn
  35:     0x7f6d286d9dac - <rustc_hir_typeck[5384bad6713105a9]::inherited::InheritedBuilder>::enter::<rustc_hir_typeck[5384bad6713105a9]::typeck_with_fallback<rustc_hir_typeck[5384bad6713105a9]::typeck::{closure#0}>::{closure#0}::{closure#1}, &rustc_middle[e87e357ae1384561]::ty::typeck_results::TypeckResults>
  36:     0x7f6d2881336a - rustc_hir_typeck[5384bad6713105a9]::typeck_with_fallback::<rustc_hir_typeck[5384bad6713105a9]::typeck::{closure#0}>
  37:     0x7f6d287399a4 - rustc_hir_typeck[5384bad6713105a9]::typeck
  38:     0x7f6d299a66f4 - rustc_query_system[a641d3afe2b8834a]::query::plumbing::try_execute_query::<rustc_query_impl[2d661b98ef5c64e8]::queries::typeck, rustc_query_impl[2d661b98ef5c64e8]::plumbing::QueryCtxt>
  39:     0x7f6d29a41b20 - rustc_query_system[a641d3afe2b8834a]::query::plumbing::get_query::<rustc_query_impl[2d661b98ef5c64e8]::queries::typeck, rustc_query_impl[2d661b98ef5c64e8]::plumbing::QueryCtxt, rustc_middle[e87e357ae1384561]::dep_graph::dep_node::DepKind>
  40:     0x7f6d297542c8 - <rustc_query_impl[2d661b98ef5c64e8]::Queries as rustc_middle[e87e357ae1384561]::ty::query::QueryEngine>::typeck
  41:     0x7f6d2877c016 - std[6161937930c6e443]::panicking::try::<(), core[a759e9f1dec68147]::panic::unwind_safe::AssertUnwindSafe<rustc_data_structures[ad989cddaa07d642]::sync::par_for_each_in<&[rustc_span[67e7a4497f54abf5]::def_id::LocalDefId], <rustc_middle[e87e357ae1384561]::hir::map::Map>::par_body_owners<rustc_hir_typeck[5384bad6713105a9]::typeck_item_bodies::{closure#0}>::{closure#0}>::{closure#0}::{closure#0}>>
  42:     0x7f6d28811395 - rustc_data_structures[ad989cddaa07d642]::sync::par_for_each_in::<&[rustc_span[67e7a4497f54abf5]::def_id::LocalDefId], <rustc_middle[e87e357ae1384561]::hir::map::Map>::par_body_owners<rustc_hir_typeck[5384bad6713105a9]::typeck_item_bodies::{closure#0}>::{closure#0}>
  43:     0x7f6d28739865 - rustc_hir_typeck[5384bad6713105a9]::typeck_item_bodies
  44:     0x7f6d2996323a - rustc_query_system[a641d3afe2b8834a]::query::plumbing::try_execute_query::<rustc_query_impl[2d661b98ef5c64e8]::queries::typeck_item_bodies, rustc_query_impl[2d661b98ef5c64e8]::plumbing::QueryCtxt>
  45:     0x7f6d29a141ef - rustc_query_system[a641d3afe2b8834a]::query::plumbing::get_query::<rustc_query_impl[2d661b98ef5c64e8]::queries::typeck_item_bodies, rustc_query_impl[2d661b98ef5c64e8]::plumbing::QueryCtxt, rustc_middle[e87e357ae1384561]::dep_graph::dep_node::DepKind>
  46:     0x7f6d29753d8e - <rustc_query_impl[2d661b98ef5c64e8]::Queries as rustc_middle[e87e357ae1384561]::ty::query::QueryEngine>::typeck_item_bodies
  47:     0x7f6d2892e4e8 - <rustc_session[29c6e5af3d515c5e]::session::Session>::time::<(), rustc_hir_analysis[2dedc18b32354586]::check_crate::{closure#7}>
  48:     0x7f6d2889bdae - rustc_hir_analysis[2dedc18b32354586]::check_crate
  49:     0x7f6d283c6e52 - rustc_interface[e7497e78aff66ef4]::passes::analysis
  50:     0x7f6d299a87ca - rustc_query_system[a641d3afe2b8834a]::query::plumbing::try_execute_query::<rustc_query_impl[2d661b98ef5c64e8]::queries::analysis, rustc_query_impl[2d661b98ef5c64e8]::plumbing::QueryCtxt>
  51:     0x7f6d29a41db1 - rustc_query_system[a641d3afe2b8834a]::query::plumbing::get_query::<rustc_query_impl[2d661b98ef5c64e8]::queries::analysis, rustc_query_impl[2d661b98ef5c64e8]::plumbing::QueryCtxt, rustc_middle[e87e357ae1384561]::dep_graph::dep_node::DepKind>
  52:     0x7f6d29735a5e - <rustc_query_impl[2d661b98ef5c64e8]::Queries as rustc_middle[e87e357ae1384561]::ty::query::QueryEngine>::analysis
  53:     0x7f6d2828fb05 - <rustc_interface[e7497e78aff66ef4]::passes::QueryContext>::enter::<rustc_driver[748794e2e144500f]::run_compiler::{closure#1}::{closure#2}::{closure#3}, core[a759e9f1dec68147]::result::Result<(), rustc_errors[aafc41f3c6a3f30f]::ErrorGuaranteed>>
  54:     0x7f6d2829103c - <rustc_interface[e7497e78aff66ef4]::interface::Compiler>::enter::<rustc_driver[748794e2e144500f]::run_compiler::{closure#1}::{closure#2}, core[a759e9f1dec68147]::result::Result<core[a759e9f1dec68147]::option::Option<rustc_interface[e7497e78aff66ef4]::queries::Linker>, rustc_errors[aafc41f3c6a3f30f]::ErrorGuaranteed>>
  55:     0x7f6d2826f966 - rustc_span[67e7a4497f54abf5]::with_source_map::<core[a759e9f1dec68147]::result::Result<(), rustc_errors[aafc41f3c6a3f30f]::ErrorGuaranteed>, rustc_interface[e7497e78aff66ef4]::interface::run_compiler<core[a759e9f1dec68147]::result::Result<(), rustc_errors[aafc41f3c6a3f30f]::ErrorGuaranteed>, rustc_driver[748794e2e144500f]::run_compiler::{closure#1}>::{closure#0}::{closure#0}>
  56:     0x7f6d282933fc - <scoped_tls[535ca24a62cbd8c4]::ScopedKey<rustc_span[67e7a4497f54abf5]::SessionGlobals>>::set::<rustc_interface[e7497e78aff66ef4]::interface::run_compiler<core[a759e9f1dec68147]::result::Result<(), rustc_errors[aafc41f3c6a3f30f]::ErrorGuaranteed>, rustc_driver[748794e2e144500f]::run_compiler::{closure#1}>::{closure#0}, core[a759e9f1dec68147]::result::Result<(), rustc_errors[aafc41f3c6a3f30f]::ErrorGuaranteed>>
  57:     0x7f6d2827c1aa - std[6161937930c6e443]::sys_common::backtrace::__rust_begin_short_backtrace::<rustc_interface[e7497e78aff66ef4]::util::run_in_thread_pool_with_globals<rustc_interface[e7497e78aff66ef4]::interface::run_compiler<core[a759e9f1dec68147]::result::Result<(), rustc_errors[aafc41f3c6a3f30f]::ErrorGuaranteed>, rustc_driver[748794e2e144500f]::run_compiler::{closure#1}>::{closure#0}, core[a759e9f1dec68147]::result::Result<(), rustc_errors[aafc41f3c6a3f30f]::ErrorGuaranteed>>::{closure#0}::{closure#0}, core[a759e9f1dec68147]::result::Result<(), rustc_errors[aafc41f3c6a3f30f]::ErrorGuaranteed>>
  58:     0x7f6d282976ee - std[6161937930c6e443]::panicking::try::<core[a759e9f1dec68147]::result::Result<(), rustc_errors[aafc41f3c6a3f30f]::ErrorGuaranteed>, core[a759e9f1dec68147]::panic::unwind_safe::AssertUnwindSafe<<std[6161937930c6e443]::thread::Builder>::spawn_unchecked_<rustc_interface[e7497e78aff66ef4]::util::run_in_thread_pool_with_globals<rustc_interface[e7497e78aff66ef4]::interface::run_compiler<core[a759e9f1dec68147]::result::Result<(), rustc_errors[aafc41f3c6a3f30f]::ErrorGuaranteed>, rustc_driver[748794e2e144500f]::run_compiler::{closure#1}>::{closure#0}, core[a759e9f1dec68147]::result::Result<(), rustc_errors[aafc41f3c6a3f30f]::ErrorGuaranteed>>::{closure#0}::{closure#0}, core[a759e9f1dec68147]::result::Result<(), rustc_errors[aafc41f3c6a3f30f]::ErrorGuaranteed>>::{closure#1}::{closure#0}>>
  59:     0x7f6d2828c646 - <<std[6161937930c6e443]::thread::Builder>::spawn_unchecked_<rustc_interface[e7497e78aff66ef4]::util::run_in_thread_pool_with_globals<rustc_interface[e7497e78aff66ef4]::interface::run_compiler<core[a759e9f1dec68147]::result::Result<(), rustc_errors[aafc41f3c6a3f30f]::ErrorGuaranteed>, rustc_driver[748794e2e144500f]::run_compiler::{closure#1}>::{closure#0}, core[a759e9f1dec68147]::result::Result<(), rustc_errors[aafc41f3c6a3f30f]::ErrorGuaranteed>>::{closure#0}::{closure#0}, core[a759e9f1dec68147]::result::Result<(), rustc_errors[aafc41f3c6a3f30f]::ErrorGuaranteed>>::{closure#1} as core[a759e9f1dec68147]::ops::function::FnOnce<()>>::call_once::{shim:vtable#0}
  60:     0x7f6d279038b8 - <alloc::boxed::Box<F,A> as core::ops::function::FnOnce<Args>>::call_once::hfe0a9abdbb746703
  61:     0x7f6d278f1877 - std::sys::unix::thread::Thread::new::thread_start::h84bfc89499a70691
  62:     0x7f6d276f78fd - <unknown>
  63:     0x7f6d27779a60 - <unknown>
  64:                0x0 - <unknown>

error: internal compiler error: unexpected panic

note: the compiler unexpectedly panicked. this is a bug.

note: we would appreciate a bug report: https://github.com/rust-lang/rust/issues/new?labels=C-bug%2C+I-ICE%2C+T-compiler&template=ice.md

note: rustc 1.69.0-dev running on x86_64-unknown-linux-gnu

note: compiler flags: -Z threads=1 -C codegen-units=1 -Z ui-testing -Z simulate-remapped-rust-src-base=/rustc/FAKE_PREFIX -Z translate-remapped-path-to-local-path=no -Z deduplicate-diagnostics=no -C strip=debuginfo -C prefer-dynamic -C rpath -C debuginfo=0

query stack during panic:
#0 [typeck] type-checking `main`
#1 [typeck_item_bodies] type-checking all item bodies
#2 [analysis] running analysis passes on this crate
end of query stack
```

</details>